File: apps/web/lib/standalone/matching.ts
import type { JobListing, JobMatch } from "@/lib/api-client";
import { profileSearchTerms } from "./cv-extract";

function norm(s: string): string {
  return s.trim().toLowerCase().replace(/\s+/g, " ");
}

function skillOverlap(
  candidateSkills: string[],
  jobSkills: string[]
): { score: number; matched: string[]; missing: string[] } {
  if (!jobSkills.length) {
    return { score: 0.75, matched: candidateSkills.slice(0, 5), missing: [] };
  }
  const matched: string[] = [];
  const missing: string[] = [];
  for (const js of jobSkills) {
    const njs = norm(js);
    const hit = candidateSkills.find(
      (cs) => norm(cs).includes(njs) || njs.includes(norm(cs))
    );
    if (hit) matched.push(hit);
    else missing.push(js);
  }
  return {
    score: matched.length / Math.max(jobSkills.length, 1),
    matched: [...new Set(matched)],
    missing,
  };
}

export function calculateMatchScore(
  profile: {
    skills: string[];
    education: Record<string, string>[];
    certifications: string[];
    experience: Record<string, string | number>[];
    keywords: string[];
    ai_summary?: Record<string, unknown>;
  },
  job: JobListing
) {
  const allJobSkills = [
    ...new Set([...(job.required_skills || []), ...(job.preferred_skills || [])]),
  ];
  const desc = job.description || "";
  const { score: skillScore, matched, missing } = skillOverlap(
    profile.skills,
    allJobSkills.length ? allJobSkills : profileSearchTerms(profile)
  );

  const domains = (profile.ai_summary?.primary_domains as string[]) || [];
  const blob = `${desc} ${job.title}`.toLowerCase();
  const domainHit = domains.some(
    (d) => d && d !== "general" && blob.includes(d.toLowerCase())
  );
  const domainBoost = domainHit ? 0.15 : 0;

  const expYears = profile.experience.reduce(
    (sum, e) => sum + Number(e.years || 1),
    0
  );
  const expScore = Math.min(expYears / 3, 1) * 0.25;
  const eduScore = profile.education.length ? 0.15 : 0.05;
  const composite = Math.min(
    skillScore * 0.45 + expScore + eduScore + domainBoost + 0.1,
    1
  );
  const matchScore = Math.round(composite * 100);

  return {
    match_score: matchScore,
    potential_score: Math.min(matchScore + 15, 98),
    matched_skills: matched,
    missing_skills: missing.slice(0, 10),
    match_explanation: {
      summary: matched.length
        ? `Matched ${matched.length} skills for ${job.title}`
        : "Limited overlap — try refining your CV keywords",
      skill_score: Math.round(skillScore * 100),
    },
  };
}

export function rankJobsForProfile(
  profile: Parameters<typeof calculateMatchScore>[0],
  jobs: JobListing[]
): JobMatch[] {
  return jobs
    .map((job) => {
      const scores = calculateMatchScore(profile, job);
      return {
        id: crypto.randomUUID(),
        ...scores,
        job,
        created_at: new Date().toISOString(),
      };
    })
    .sort((a, b) => b.match_score - a.match_score);
}

export function aggregateSkillGaps(matches: JobMatch[]) {
  const counts: Record<string, number> = {};
  for (const m of matches) {
    for (const s of m.missing_skills || []) {
      counts[s] = (counts[s] || 0) + 1;
    }
  }
  return Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 15)
    .map(([skill_name, count]) => ({
      id: crypto.randomUUID(),
      skill_name,
      priority: Math.min(5, Math.ceil(count / 2)),
      learning_path: {
        resources: [
          `https://www.coursera.org/search?query=${encodeURIComponent(skill_name)}`,
          `https://www.youtube.com/results?search_query=learn+${encodeURIComponent(skill_name)}`,
        ],
        estimated_weeks: 4,
      },
    }));
}

File: apps/api/app/services/matching.py
from __future__ import annotations

import math
import re
from typing import Any

from rapidfuzz import fuzz


def _normalize_skill(skill: str) -> str:
    return re.sub(r"\s+", " ", skill.strip().lower())


def _skill_overlap(candidate_skills: list[str], job_skills: list[str]) -> tuple[float, list[str], list[str]]:
    if not job_skills:
        return 0.75, candidate_skills[:5], []

    norm_candidate = {_normalize_skill(s): s for s in candidate_skills}
    matched: list[str] = []
    missing: list[str] = []

    for job_skill in job_skills:
        norm_job = _normalize_skill(job_skill)
        best_ratio = 0
        best_match = None
        for norm_c, original in norm_candidate.items():
            ratio = fuzz.token_set_ratio(norm_job, norm_c) / 100
            if ratio > best_ratio:
                best_ratio = ratio
                best_match = original
        if best_ratio >= 0.75 and best_match:
            if best_match not in matched:
                matched.append(best_match)
        else:
            missing.append(job_skill)

    score = len(matched) / max(len(job_skills), 1)
    return score, matched, missing


def _experience_score(experience: list[dict[str, Any]], description: str) -> float:
    years = 0.0
    for exp in experience:
        years += float(exp.get("years", 0) or 0)
    if years == 0 and experience:
        years = len(experience) * 1.5

    desc = description.lower()
    if any(k in desc for k in ("senior", "lead", "principal")):
        required = 5
    elif any(k in desc for k in ("mid", "intermediate")):
        required = 3
    elif any(k in desc for k in ("junior", "entry", "graduate", "intern")):
        required = 1
    else:
        required = 2

    return min(years / required, 1.0) if required else 0.5


def _education_score(education: list[dict[str, Any]], description: str) -> float:
    if not education:
        return 0.3
    desc = description.lower()
    levels = {"phd": 1.0, "master": 0.9, "bachelor": 0.8, "diploma": 0.6, "certificate": 0.5}
    best = 0.5
    for edu in education:
        degree = str(edu.get("degree", "")).lower()
        for key, val in levels.items():
            if key in degree:
                best = max(best, val)
    if "degree" in desc or "bachelor" in desc:
        return min(best, 1.0)
    return best


def _certification_score(certifications: list[str], job_skills: list[str]) -> float:
    if not certifications:
        return 0.2
    norm_certs = {_normalize_skill(c) for c in certifications}
    hits = sum(1 for s in job_skills if _normalize_skill(s) in norm_certs)
    return min(hits / max(len(job_skills), 1), 1.0) if job_skills else 0.5


def _keyword_score(keywords: list[str], description: str) -> float:
    if not keywords:
        return 0.3
    desc = description.lower()
    hits = sum(1 for k in keywords if k.lower() in desc)
    return min(hits / max(len(keywords), 1), 1.0)


def calculate_match_score(
    profile: dict[str, Any],
    job: dict[str, Any],
) -> dict[str, Any]:
    skills = profile.get("skills", [])
    education = profile.get("education", [])
    certifications = profile.get("certifications", [])
    experience = profile.get("experience", [])
    keywords = profile.get("keywords", [])

    required = job.get("required_skills", []) or []
    preferred = job.get("preferred_skills", []) or []
    all_job_skills = list(dict.fromkeys(required + preferred))
    description = job.get("description", "")

    skill_score, matched, missing = _skill_overlap(skills, required or all_job_skills)
    exp_score = _experience_score(experience, description)
    edu_score = _education_score(education, description)
    cert_score = _certification_score(certifications, all_job_skills)
    kw_score = _keyword_score(keywords, description)
    domain_terms = [str(d).lower() for d in profile.get("primary_domains", [])]
    domain_terms += [str(d).lower() for d in profile.get("ai_summary", {}).get("domains", [])]
    if domain_terms:
        desc_lower = description.lower()
        title_lower = job.get("title", "").lower()
        domain_hits = sum(
            1 for domain in set(domain_terms)
            if domain and domain != "general" and (domain in desc_lower or domain in title_lower)
        )
        kw_score = max(kw_score, min(domain_hits / max(len(set(domain_terms)), 1), 1.0))

    composite = (
        skill_score * 0.40
        + exp_score * 0.25
        + edu_score * 0.15
        + cert_score * 0.10
        + kw_score * 0.10
    )
    match_score = int(round(composite * 100))

    # Potential score if top 3 missing skills acquired
    simulated_skills = skills + missing[:3]
    sim_score, _, _ = _skill_overlap(simulated_skills, required or all_job_skills)
    potential = int(
        round(
            (
                sim_score * 0.40
                + exp_score * 0.25
                + edu_score * 0.15
                + cert_score * 0.10
                + kw_score * 0.10
            )
            * 100
        )
    )

    explanation = {
        "summary": f"Strong alignment on {len(matched)} core skills" if matched else "Limited skill overlap",
        "skill_score": round(skill_score * 100),
        "experience_score": round(exp_score * 100),
        "education_score": round(edu_score * 100),
        "certification_score": round(cert_score * 100),
        "keyword_score": round(kw_score * 100),
    }

    return {
        "match_score": match_score,
        "potential_score": max(potential, match_score),
        "matched_skills": matched,
        "missing_skills": missing[:10],
        "match_explanation": explanation,
    }


def aggregate_skill_gaps(matches: list[dict[str, Any]]) -> list[dict[str, Any]]:
    counts: dict[str, int] = {}
    for match in matches:
        for skill in match.get("missing_skills", []):
            counts[skill] = counts.get(skill, 0) + 1

    ranked = sorted(counts.items(), key=lambda x: x[1], reverse=True)
    return [
        {
            "skill_name": skill,
            "priority": min(5, math.ceil(count / 2)),
            "learning_path": {
                "resources": [
                    f"https://www.coursera.org/search?query={skill.replace(' ', '%20')}",
                    f"https://www.youtube.com/results?search_query=learn+{skill.replace(' ', '+')}",
                ],
                "estimated_weeks": 4,
            },
        }
        for skill, count in ranked[:15]
    ]
